ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/RemoteDataAugmenter.java

```diff
@@ -0,0 +1,157 @@
+import io.confluent.ksql.services.SimpleKsqlClient;
+import io.confluent.ksql.util.KsqlRequestConfig;
+import io.confluent.ksql.util.Pair;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import org.apache.kafka.streams.state.HostInfo;
+
+
+public final class RemoteDataAugmenter {
+  private static final int TIMEOUT_SECONDS = 10;
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteDataAugmenter.class);
+
+  private final String statementText;
+  private final SessionProperties sessionProperties;
+  private final KsqlExecutionContext executionContext;
+  private final SimpleKsqlClient ksqlClient;
+
+
+  private RemoteDataAugmenter(
+      final String statementText,
+      final SessionProperties sessionProperties,
+      final KsqlExecutionContext executionContext,
+      final SimpleKsqlClient ksqlClient
+  ) {
+    this.statementText = Objects.requireNonNull(statementText);
+    this.sessionProperties = Objects.requireNonNull(sessionProperties);
+    this.executionContext = Objects.requireNonNull(executionContext);
+    this.ksqlClient = Objects.requireNonNull(ksqlClient);
+    ;
+  }
+
+  public static RemoteDataAugmenter create(
+      final String statementText,
+      final SessionProperties sessionProperties,
+      final KsqlExecutionContext executionContext,
+      final SimpleKsqlClient ksqlClient
+  ) {
+    return new RemoteDataAugmenter(
+        statementText, sessionProperties, executionContext, ksqlClient);
+  }
+
+
+  public <R> R augmentWithRemote(
+      final R localResult,
+      final BiFunction<R, Pair<List<KsqlEntity>, Set<HostInfo>>, R> mergeFunc
+  ) {
+    Objects.requireNonNull(mergeFunc);
+    if (sessionProperties.getInternalRequest()) {
+      return mergeFunc.apply(localResult, new Pair<>(ImmutableList.of(), ImmutableSet.of()));
+    }
+    final Pair<List<KsqlEntity>, Set<HostInfo>> remoteResults = fetchAllRemoteResults();
+
+    return mergeFunc.apply(localResult, remoteResults);
+  }
+
+  private CompletableFuture<RestResponse<KsqlEntityList>> fetchRemoteData(
+      final String statementText,
+      final HostInfo host,
+      final Executor executor
+  ) {
+    return CompletableFuture.supplyAsync(() -> ksqlClient.makeKsqlRequest(
+        ServerUtil.buildRemoteUri(
+            sessionProperties.getLocalUrl(),
+            host.host(),
+            host.port()
+        ),
+        statementText,
+        Collections.singletonMap(KsqlRequestConfig.KSQL_REQUEST_INTERNAL_REQUEST, true)),
+        executor);
```

Comment:
extreme nit: can you format the `supplyAsync` function in a more readable way? I had to do a double take and check if `makeKsqlRequest()` takes an executor or not since you'd have to count the parentheses carefully to see the the request function doesn't the executor and that the executor is the second argument of `supplyAsync()`

Reply:
Done!